### PR TITLE
Add preshared_key variable to be added to systemd-networkd config.

### LINKED ISSUE #46: Add pre-shared key per peer
Wireguard can use an optional pre-shared key per peer [1], in order to increase safety. This has not yet been implemented in this module. Can you add this parameter?

[1] https://git.zx2c4.com/wireguard-tools/about/src/man/wg.8

### PR DESCRIPTION
**Pull request by jhammons**
#### Pull Request (PR) description
Add preshared_key variable to be added to systemd-networkd config.

#### This Pull Request (PR) fixes the following issues
Fixes #46 .

### CODE REVIEW

File: templates/netdev.epp [hunk 1]
```diff
@@ -2,6 +2,7 @@
       Stdlib::Port $dport,
       Wireguard::Peers $peers,
       Optional[String] $description,
+      Optional[String] $preshared_key,
```
Comment (root-expert) [inline templates/netdev.epp:5]:
```suggestion
      Optional[String[1]] $preshared_key,
```

File: templates/netdev.epp [hunk 2]
```diff
@@ -18,6 +19,7 @@ MTUBytes=<%= $mtu %>
 
 [WireGuard]
 PrivateKeyFile=/etc/wireguard/<%= $interface %>
+
```
Comment (root-expert) [inline templates/netdev.epp:22]: New line seems unecessary here :smile:

**Comment by smortex** [comment]
Can you please add a test in `spec/defines/interface_spec.rb` that proves this has the expected behavior? Thanks!

**Review by root-expert — COMMENTED** [review]
Thanks for the PR! Can you please add a unit spec for the new parameter?

**Comment by Southparkfan** [comment]
Folks, anything that can help this get merged? I have a patch (written by myself, AGPL-3.0 is fine) that includes a basic test to confirm the presence of a PSK.
```diff
diff --git a/spec/defines/interface_spec.rb b/spec/defines/interface_spec.rb
index 5712de2..97d9c02 100644
--- a/spec/defines/interface_spec.rb
+++ b/spec/defines/interface_spec.rb
@@ -122,7 +122,7 @@ describe 'wireguard::interface', type: :define do
         it { is_expected.to contain_ferm__rule("allow_wg_#{title}") }
       end

-      context 'with required params and without firewall rules and with configured addresses' do
+      context 'with required params and without firewall rules and with configured addresses and with preshared key' do
         let :params do
           {
             public_key: 'blabla==',
@@ -132,6 +132,7 @@ describe 'wireguard::interface', type: :define do
             # that would configure IPv4+IPv6, but GHA doesn't provide IPv6 for us
             destination_addresses: [facts[:networking]['ip'],],
             addresses: [{ 'Address' => '192.168.218.87/32', 'Peer' => '172.20.53.97/32' }, { 'Address' => 'fe80::ade1/64', },],
+            preshared_key: 'mypsk==',
           }
         end

@@ -147,6 +148,7 @@ describe 'wireguard::interface', type: :define do
         it { is_expected.to contain_file("/etc/systemd/network/#{title}.network").with_content(%r{Address=192.168.218.87/32}) }
         it { is_expected.to contain_file("/etc/systemd/network/#{title}.network").with_content(%r{Peer=172.20.53.97/32}) }
         it { is_expected.to contain_file("/etc/systemd/network/#{title}.network").with_content(%r{Address=fe80::ade1/64}) }
+        it { is_expected.to contain_file("/etc/systemd/network/#{title}.netdev").with_content(%r{PresharedKey=mypsk==}) }
         it { is_expected.not_to contain_ferm__rule("allow_wg_#{title}") }
       end

diff --git a/templates/netdev.epp b/templates/netdev.epp
index d4f518c..9f6ce7b 100644
--- a/templates/netdev.epp
+++ b/templates/netdev.epp
@@ -3,6 +3,7 @@
       Wireguard::Peers $peers,
       Optional[String] $description,
       Optional[Integer] $mtu,
+      Optional[String[1]] $preshared_key,
 | -%>
 # THIS FILE IS MANAGED BY PUPPET
 # based on https://dn42.dev/howto/wireguard
@@ -26,6 +27,9 @@ ListenPort=<%= $dport %>
 Description=<%= $peer['description'] %>
 <% } -%>
 PublicKey=<%= $peer['public_key'] %>
+<% if $preshared_key { -%>
+PresharedKey=<%= $preshared_key %>
+<% } -%>
 <% if $peer['endpoint'] { -%>
 Endpoint=<%= $peer['endpoint'] %>
 <% } -%>
```

Not an experienced user of rspec, but it passed the rake tests (well, I had to add `PUPPET_GEM_VERSION="~> 7,17"` to my `Dockerfile`) locally. Would be awesome if we can merge this together with the code in this PR.

**Comment by root-expert** [comment]
Hello @Southparkfan, you can open a new PR containing the changes and if everything's ok we can merge it. :smile:

**Comment by Southparkfan** [comment]
@root-expert: new PR is #49.

However, I didn't notice that @jhammons marked this PR as 'draft' a few hours ago. Jason, do you wish to amend my change into your PR, are you going to close this PR in favor of the other one, ...?

**Comment by jhammons** [comment]
Closing in favor of https://github.com/voxpupuli/puppet-wireguard/pull/49.

**Pull request closed without merging**